Psychtoolbox/PsychHardware/EyelinkToolbox/EyelinkDemos/SR-ResearchDemos/MRI_BlockRecord/EyeLink_MRI_BlockRecord.m

```diff
@@ -0,0 +1,430 @@
+            stimArray = Screen('GetImage', window, [], 'drawBuffer'); % Copy backbuffer to be used as stimulus
+            imwrite(stimArray, stimName); % Save .bmp stimulus file in experment folder           
+            % Convert stimulus to texture so it is ready for drawing later on
+            stimTexture = Screen('MakeTexture', window, stimArray); % Convert to texture
+            
+            % STEP 7.2: START TRIAL AND DRAW CROSSHAIRS FOR DATAVIEWER
+            
+            if i ~= 1 % If not first trial in a block
+                % Write TRIALID message to EDF file: marks the start of a trial for DataViewer
+                % TRIALID message for first trial in the block was written before recording began
+                % See DataViewer manual section: Protocol for EyeLink Data to Viewer Integration > Defining the Start and End of a Trial
+                trialCount = trialCount + 1; % Add 1 to trial counter
+                Eyelink('Message', 'TRIALID %d', trialCount);
+            end
+            
+            % Write messages to EDF: prepare backdrop and draw central crosshairs for DataViewer
+            % See DataViewer manual section: Protocol for EyeLink Data to Viewer Integration > Simple Drawing
+            Eyelink('Message', '!V CLEAR %d %d %d', el.backgroundcolour(1), el.backgroundcolour(2), el.backgroundcolour(3));
+            Eyelink('Message', '!V DRAWLINE 0 0 0 %d %d %d %d', round(width/2-20), round(height/2), round(width/2+20), round(height/2));
+            Eyelink('Message', '!V DRAWLINE 0 0 0 %d %d %d %d', round(width/2), round(height/2-20), round(width/2), round(height/2+20));
+            
+            % STEP 7.3: WAIT FOR MRI TRIGGER AND PRESENT STIMULUS; CREATE DATAVIEWER STIMULUS BACKDROP AND INTEREST AREA
+            
+            % Present initial trial image
+            Screen('DrawTexture', window, stimTexture); % Prepare stimulus texture on backbuffer            
+            % Wait for MRI trigger (keyboard key "t" in this demo)
+            if i == 1 % If first trial in a block
+                KbReleaseWait; % Wait until user releases keys on keyboard:
+                keyTrigger = KbName('t');% Identify key code for keyboard key "t"
+                while 1
+                    [~, ~, keyCode] = KbCheck;
+                    if keyCode(keyTrigger)
+                        % Write message to EDF file to mark the time when the trigger is received
+                        Eyelink('Message', 'TRIGGER_RECEIVED');
+                        blockOnset = GetSecs; % Block onset time
+                        vbl = Screen('Flip', window); % Present stimulus
+                        break;
+                    end
+                end
+            else % All subsequent trials in block
+                vbl = Screen('Flip', window, blockOnset + (trialDur*(i-1)) - 0.5*ifi); % Present stimulus. Allow half flip interval for precise flip timing)

```

Comment:
As blockOnset is not a time returned from Screen('Flip'), but from GetSecs in response to scanner triggers, blockOnset is not a time corresponding to start of a video refresh cycle, so this math will go wrong. Setting blockOnset = vbl; in the if-clause above would be more robust.

More importantly: The formula for defining the desired stimulus onset time only makes sense if trialDur() is specified in video refresh cycles, e.g., 60 for one second on a 60 Hz monitor. But trialDur is in seconds, so the `-0.5 * ifi` part of the math makes not so much sense. Dropping it may not change anything, or even make it more accurate, and less confusing.

Reply:
will include as priority item for future revision alongside the other suggested changes to the MRI demo using triggerDevice to improve multiple HID support